core/controller/src/main/scala/whisk/core/controller/GatewayRoutesAPI.scala

```diff
@@ -0,0 +1,191 @@
+                  p trySuccess result.entity.asString
+                }
+                case Failure(error) => {
+                  p tryFailure error
+                }
+             })
+             p.future
+       }
+       /**
+        ** Handles invoking get actions
+        **
+        * @return Future[String] from action result
+        */
+        def invokeActionGet(requestBody: String ) : Future[String]  = {
+            val actionPath = Path("/api/"+apiversion) / "namespaces" / "whisk.system" / "actions" / "routemgmt" / "getApi"
+            val pipeline: HttpRequest => Future[HttpResponse] = (
+                addCredentials(validCredentials)
+                    ~> sendReceive
+            )
+            val p = Promise[String]()
+            val response: Future[HttpResponse] = pipeline(Post(url.withPath(actionPath).toString.concat("?blocking=true"), HttpEntity(ContentTypes.`application/json`, requestBody))  )
+            response.onComplete({
+                case Success(result: HttpResponse) => {
+                    p trySuccess result.entity.asString
+                }
+                case Failure(error) => {
+                    p tryFailure error
+                }
+            })
+            p.future
+        }
+        /**
+        * Handles invoking delete action
+        **
+        * @return Future[String] from action result
+        */
+        def invokeActionDelete(requestBody: String ) : Future[String]  = {
+            val actionPath = Path("/api/"+apiversion) / "namespaces" / "whisk.system" / "actions" / "routemgmt" / "deleteApi"
+            val pipeline: HttpRequest => Future[HttpResponse] = (
+                    addCredentials(validCredentials)
+                             ~> sendReceive
+            )
+            val p = Promise[String]()
+            val response: Future[HttpResponse] = pipeline(Post(url.withPath(actionPath).toString.concat("?blocking=true"), HttpEntity(ContentTypes.`application/json`, requestBody))  )
+            response.onComplete({
+                case Success(result: HttpResponse) => {
+                   p trySuccess result.entity.asString
+                }
+                case Failure(error) => {
+                   p tryFailure error
+                }
+            })
+            p.future
+      }
+      /**
+      * Handles the controller's api gateway routes
+      */
+      def routes(user: Identity)(implicit transid: TransactionId) = {
+       lazy val collectionPrefix = pathPrefix( (EntityPath.DEFAULT.toString.r | Segment) / "routes")
+       get {
+         collectionPrefix { segment => {
+           entity(as[String]) { pattern =>
+             parameters("basepath"?"", "operation"?"","relpath"?"") { (basepath, operation,relpath) =>
+               var postBody = "{" + "\"namespace\":" +"\"" + segment + "\"," + "\"basepath\":" +"\"" + basepath + "\"," + "\"operation\":" +"\"" + operation + "\"," + "\"relpath\":" +"\"" + relpath + "\"}";
```

Comment:
Also: drop the `var`. There are several occurences of checks against empty `String`s. From how I read the code, these should be modeled as `Option[String]` instead. Would that make more sense?